src/emu/emuopts.cpp

```diff
@@ -213,133 +218,777 @@ const options_entry emu_options::s_option_entries[] =
+			if (iter != m_vec.end())
+				m_vec.erase(iter);
+		}
+
+		std::vector<const std::string *>::iterator begin() { return m_vec.begin(); }
+		std::vector<const std::string *>::iterator end() { return m_vec.end(); }
+
+	private:
+		std::vector<const std::string *> m_vec;
+	};
+
+
+	//-------------------------------------------------
+	//  get_full_option_names
+	//-------------------------------------------------
+
+	std::vector<std::string> get_full_option_names(const device_image_interface &image)
+	{
+		std::vector<std::string> result;
+
+		result.push_back(image.instance_name());
+		result.push_back(image.brief_instance_name());
+
+		if (strcmp(image.device_typename(image.image_type()), image.instance_name().c_str()) == 0)
+		{
+			result.push_back(image.instance_name() + "1");
+			result.push_back(image.brief_instance_name() + "1");
+		}
+		return result;
+	}
+}
+
+
+//**************************************************************************
 //  EMU OPTIONS
 //**************************************************************************
 
 //-------------------------------------------------
 //  emu_options - constructor
 //-------------------------------------------------
 
-emu_options::emu_options()
-: core_options()
-, m_coin_impulse(0)
-, m_joystick_contradictory(false)
-, m_sleep(true)
-, m_refresh_speed(false)
-, m_ui(UI_CABINET)
+emu_options::emu_options(bool general_only)
+	: m_system(nullptr)
+	, m_coin_impulse(0)
+	, m_joystick_contradictory(false)
+	, m_sleep(true)
+	, m_refresh_speed(false)
+	, m_ui(UI_CABINET)
 {
+	// add entries
+	if (!general_only)
+	{
+		add_entry(std::make_unique<system_name_option_entry>(*this));
+		add_entry(std::make_unique<software_name_option_entry>(*this));
+	}
 	add_entries(emu_options::s_option_entries);
+
+	// adding handlers to keep copies of frequently requested options in member variables
+	set_value_changed_handler(OPTION_COIN_IMPULSE,				[this](const char *value) { m_coin_impulse = int_value(OPTION_COIN_IMPULSE); });
+	set_value_changed_handler(OPTION_JOYSTICK_CONTRADICTORY,	[this](const char *value) { m_joystick_contradictory = bool_value(OPTION_JOYSTICK_CONTRADICTORY); });
+	set_value_changed_handler(OPTION_SLEEP,						[this](const char *value) { m_sleep = bool_value(OPTION_SLEEP); });
+	set_value_changed_handler(OPTION_REFRESHSPEED,				[this](const char *value) { m_sleep = bool_value(OPTION_REFRESHSPEED); });
+	set_value_changed_handler(OPTION_UI, [this](const std::string &value)
+	{
+		if (value == "simple")
+			m_ui = UI_SIMPLE;
+		else
+			m_ui = UI_CABINET;
+	});
+}
+
+
+//-------------------------------------------------
+//  emu_options - destructor
+//-------------------------------------------------
+
+emu_options::~emu_options()
+{
+}
+
+
+//-------------------------------------------------
+//	system_name
+//-------------------------------------------------
+
+const char *emu_options::system_name() const
+{
+	return m_system ? m_system->name : "";
```

Comment:
Existing usages of emu_options::system_name() expect an empty string.  An example: cli_frontend::execute_commands()